lib/azure/armrest/virtual_machine_service.rb

```diff
@@ -121,12 +121,169 @@ def stop(vmname, group = configuration.resource_group)
+
+        vm = get(vmname, vmgroup)
+
+        delete_and_wait(self, vmname, vmgroup, options)
+
+        # Must delete network interfaces first if you want to delete
+        # IP addresses or network security groups.
+        if options[:network_interfaces] || options[:ip_addresses] || options[:network_security_groups]
+          delete_associated_nics(vm, options)
+        end
+
+        if options[:os_disk] || options[:storage_account]
+          delete_associated_disk(vm, options)
+        end
+      end
+
       def model_class
         VirtualMachineModel
       end
 
       private
 
+      # Deletes any NIC's associated with the VM, and optionally any public IP addresses
+      # and network security groups.
+      #
+      def delete_associated_nics(vm, options)
+        nis = Azure::Armrest::Network::NetworkInterfaceService.new(configuration)
+        nics = vm.properties.network_profile.network_interfaces.map(&:id)
+
+        nics.each do |nic_id_string|
+          nic = get_associated_resource(nic_id_string)
+          delete_and_wait(nis, nic.name, nic.resource_group, options)
+
+          if options[:ip_addresses]
+            ips = Azure::Armrest::Network::IpAddressService.new(configuration)
+            nic.properties.ip_configurations.each do |ip|
+              ip = get_associated_resource(ip.properties.public_ip_address.id)
+              delete_and_wait(ips, ip.name, ip.resource_group, options)
+            end
+          end
+
+          if options[:network_security_groups]
+            nsgs = Azure::Armrest::Network::NetworkSecurityGroupService.new(configuration)
+            nic.properties.network_security_group
+            nsg = get_associated_resource(nic.properties.network_security_group.id)
+            delete_and_wait(nsgs, nsg.name, nsg.resource_group, options)
+          end
+        end
+      end
+
+      # This deletes the OS disk from the storage account that's backing the
+      # virtual machine, along with the .status file. This does NOT delete
+      # copies of the disk.
+      #
+      # If the option to delete the entire storage account was selected, then
+      # it will not bother with deleting invidual files from the storage
+      # account first.
+      #
+      def delete_associated_disk(vm, options)
+        uri = Addressable::URI.parse(vm.properties.storage_profile.os_disk.vhd.uri)
+
+        # The uri looks like https://foo123.blob.core.windows.net/vhds/something123.vhd
+        storage_acct_name = uri.host.split('.').first     # storage name, e.g. 'foo123'
+        storage_acct_disk = File.basename(uri.to_s)       # disk name, e.g. 'something123.vhd'
+        storage_acct_path = File.dirname(uri.path)[1..-1] # container, e.g. 'vhds'
+
+        # Must find it this way because the resource group information isn't provided
+        sas = Azure::Armrest::StorageAccountService.new(configuration)
+        storage_acct_obj  = sas.list_all.find { |s| s.name == storage_acct_name }
```

Comment:
I'll refactor this to look in the VM's resource group first, and only resort to the list_all if the first call doesn't work. The storage account will *usually* be in the same resource group.

Another option would be to cache those sas.list_all results.